Mathlib/Analysis/Normed/Group/Ultra.lean

```diff
@@ -0,0 +1,175 @@
+    ‖x ^ n‖₊ ≤ ‖x‖₊ := by
+  induction n with
+  | zero => simp
+  | succ n hn => simpa [pow_add, hn] using nnnorm_mul_le_max (x ^ n) x
+
+@[to_additive]
+lemma norm_pow_le (x : S) (n : ℕ) :
+    ‖x ^ n‖ ≤ ‖x‖ :=
+  nnnorm_pow_le x n
+
+@[to_additive]
+lemma nnnorm_zpow_le (x : S) (z : ℤ) :
+    ‖x ^ z‖₊ ≤ ‖x‖₊ := by
+  cases z <;>
+  simpa using nnnorm_pow_le _ _
+
+@[to_additive]
+lemma norm_zpow_le (x : S) (z : ℤ) :
+    ‖x ^ z‖ ≤ ‖x‖ :=
+  nnnorm_zpow_le x z
+
+end Group
+
+section CommGroup
+
+variable {M ι : Type*} [SeminormedCommGroup M] [IsUltrametricDist M]
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product.
+This version is phrased using `Finset.sup'` and `Finset.Nonempty` due to `Finset.sup`
+operating over an `OrderBot`, which `ℝ` is not.
+-/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum.
+This version is phrased using `Finset.sup'` and `Finset.Nonempty` due to `Finset.sup`
+operating over an `OrderBot`, which `ℝ` is not. "]
+lemma _root_.Finset.Nonempty.norm_prod_le_sup'_norm {s : Finset ι} (hs : s.Nonempty) (f : ι → M) :
+    ‖∏ i ∈ s, f i‖ ≤ s.sup' hs (‖f ·‖) := by
+  simp only [Finset.le_sup'_iff]
+  refine Finset.Nonempty.cons_induction ?_ ?_ hs
+  · simp
+  rintro a s ha _ ⟨b, hb, IH⟩
+  simp only [Finset.prod_cons, Finset.mem_cons, exists_eq_or_imp]
+  refine (le_total ‖∏ i ∈ s, f i‖ ‖f a‖).imp ?_ ?_ <;> intro h
+  · exact (norm_mul_le_max _ _).trans (by simp [h])
+  · exact ⟨b, hb, (norm_mul_le_max _ _).trans (by simpa [h] using IH)⟩
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product. -/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum."]
+lemma _root_.Finset.nnnorm_prod_le_sup_nnnorm (s : Finset ι) (f : ι → M) :
+    ‖∏ i ∈ s, f i‖₊ ≤ s.sup (‖f ·‖₊) := by
+  rcases s.eq_empty_or_nonempty with rfl|hs
+  · simp
+  rw [← Finset.sup'_eq_sup hs, ← NNReal.coe_le_coe,
+    Finset.comp_sup'_eq_sup'_comp hs toReal coe_mono.map_sup]
+  simpa only [coe_nnnorm', Function.comp_apply, Finset.le_sup'_iff] using
+    hs.norm_prod_le_sup'_norm f
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product. -/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum."]
+lemma _root_.Fintype.nnnorm_prod_le_sup_univ_norm [Fintype ι] (f : ι → M) :
+    ‖∏ i, f i‖₊ ≤ Finset.univ.sup (‖f ·‖₊) := by
+  simpa using Finset.univ.nnnorm_prod_le_sup_nnnorm f
+
+/-- Nonarchimedean norm of a product is less than or equal the norm of any term in the product. -/
+@[to_additive "Nonarchimedean norm of a sum is less than or equal the norm of any term in the sum."]
+lemma _root_.Fintype.norm_prod_le_sup'_univ_norm [Nonempty ι] [Fintype ι] (f : ι → M) :
+    ‖∏ i, f i‖ ≤ Finset.univ.sup' Finset.univ_nonempty (‖f ·‖) := by
+  simpa using Finset.univ_nonempty.norm_prod_le_sup'_norm f
+
```

Comment:
```suggestion

/--
Generalised ultrametric triangle inequality for nonempty finite products in groups with an
ultrametric norm.
-/
@[to_additive "Generalised ultrametric triangle inequality for nonempty finite sums in additive
groups with an ultrametric norm ."]
lemma norm_prod_le_of_forall_le_of_nonempty {s : Finset ι} (hs : s.Nonempty) {f : ι → M} {C : ℝ}
    (hC : ∀ i ∈ s, ‖f i‖ ≤ C) : ‖∏ i ∈ s, f i‖ ≤ C :=
  (hs.norm_prod_le_sup'_norm f).trans (Finset.sup'_le hs _ hC)

/--
Generalised ultrametric triangle inequality for finite products in groups with an ultrametric
norm.
-/
@[to_additive "Generalised ultrametric triangle inequality for finite sums in additive groups
with an ultrametric norm."]
lemma norm_prod_le_of_forall_le_of_nonneg {s : Finset ι} {f : ι → M} {C : ℝ}
    (h_nonneg : 0 ≤ C) (hC : ∀ i ∈ s, ‖f i‖ ≤ C) : ‖∏ i ∈ s, f i‖ ≤ C := by
  lift C to NNReal using h_nonneg
  exact (s.nnnorm_prod_le_sup_nnnorm f).trans (Finset.sup_le <| mod_cast hC)
  
```

Reply:
(feel free to ignore this if you prefer and I'll keep these lemmas in the other PR instead).